### ISSUE TITLE
Suggestions and thoughts

### ISSUE THREAD

**Issue by chryzsh** [archi-contribs/risk-security#1]
As far as I have thought about this, I realize Thorbjørn's vision is something like a web dashboard for inspecting models and their attributes in terms of security. I have started looking at some of the possibilities for such a solution based on the Archimate Modelling Tool. I may be jumping too far ahead, so don't take this too seriously. These are my thoughts:

*The existing export to HTML tool is decent, but is lacking a lot of the features we are imagining.
*We want to have increased focus on attributes of model elements and relations, with easier handling of them in the Archimatetool.
*Attributes must be reflected in both the tool and in whatever front-end solution we choose.
*A front-end solution must be developed, unless using the HTML export tool.
*Front-end must be deployed either way, thus requiring a back-end of some sort.
*The front-end must work as a presentation with the possibility of queries, thus perhaps requiring a database.
*Archi has a guide for developing import/export tools. Hooray! http://www.archimatetool.com/dev/import-export
*Must scale for multiple, large models, and easy deployment of new models must be possible.
*A back-end is required for deploying (securely).

Some different possibilites, not very detailed or anything:
* Fork Archimatetool and develop desired functionality, while using existing tools for exporting to HTML and XML. This will require redeveloping the export tools as well. This will be time consuming and requires great Java comprehension. Advantages are less challenges with the handling of XML, XSLT and such.

*Using exported XML and creating a new front end with HTML, Javascript and XSLT. May not scale well unless imported to database and handled correctly. Using XSLT is pretty new to me and I don't know if handling the XML is that easy due to the high complexity of it in this case.

* Using exported XML, parse to Java objects in a collection of some sort. Then building both a backend using Spring and/or Hibernate and a front-end using HTML and Javascript.

**Comment by chryzsh** [comment]
**Further thoughts after discussion with @thellef.**
We want to split the solution into two more or less independent parts.
* The first is a fork of Archimatetool with functionality for attributes based on the taxonomy of the Open Group Standard, Risk Taxonomy (O-RT), Version 2.0
* The second is a tool using XSLT for transforming the XML (The Open Group ArchiMate Model Exchange File Format) to something more readable. Perhaps providing a front-end made up of HTML and Javascript for presenting the data in a dashboard style

**Some additional points:**
* IDU will provide a back end when necessary.
* The JSON idea is on hold at the moment
* Developers for Java will be provided
* I am looking into XSLT in a separate repository

**Comment by jbsarrodie** [comment]
Hi,

Some remarks...

I don't think forking Archi is the way to go (other tried... and failed) because this would mean that your product would be based on a fixed in time version and so a lot of backport work to include upcomming updates (ArchiMate 3 for example). Creating a plugin or an extension is really a better way. You have all needed documentation available for an import/export plugin. If more than that is needed (more UI integration) an extension (more generic eclipse plugin) is a good solution as you'll then have more freedom on UI stuff. I haven't done such work for the moment, but Phil did it several months ago for the "Hammer" plugin (aka "Validator") which is now included mainstream but was in fact created as an extension.

The new report plugin is in fact based on a generic template engine. This mean that you can very easily change template files to suit you needs. I even thought about creating a more generic export/report engine based on it in which user could select whatever templates he want. Doing it this way would allow you to generate the exact format you want (custom XML, JSON, SQL script...).

FYI, I'm thinking about creating a script/query engine for Archi. If read-only, then the template based approch which generate a small web application containing model loaded in a [JS based RDBMS](http://alasql.org/) is my way to go. If read-write is needed, then some new model query/management/update JAVA classes used through a (JAVA based) JS interpreter console is my way to go.

Last, I know some solutions to export diagrams as DIV objects (with right position, class reflected archimate classe, name and properties as attributes...) instead of one big bitmap. Not fully tested but should be useable though.

**Comment by chryzsh** [comment]
You are absolutely right about making a plugin or extensions. However, I don't know the difference between a plugin and an extension in this context. Maybe you can elaborate on that? We are requiring a lot of UI integration, therefore I didn't see plugin as a viable option at first.

About the report plugin. Changing the template files is a possibility, but I don't know if patching on that will suit our needs. However, this won't require us to learn XSLT. I will definitely look into it.

The JS based RDBMS model may be very interesting. Let me know if you get anywhere with that.

**Comment by jbsarrodie** [comment]
Hi,

Just to make it clear, from an Eclipse point of view, there'are only things named plugins, so in the following, I will assume:
 * **plugin** : Eclipse plugin based on Archi import/export example. Only impact on UI is addition of menu entries inside File>Import and File>Export.
 * **extension** : Eclipse plugin using any potential Eclipse Framework solution to hook into the UI. A good example is the "Validator" ([com.archimatetool.hammer](https://github.com/archimatetool/archi/tree/master/com.archimatetool.hammer)). If you look at its plugin.xml file (using Eclipse), you'll see on the "Extension" tab that many extension points are used in order to add menu entries, icons in toolbar, entry in the context menu associated with the model tree...

In fact, using this definition, you'll see that many Archi features are developed as extensions (Validator, Visualiser, HTML reports, Jasper Reports...), the point is just that they are shipped by default with Archi ;-). So there's nothing you **can't** do using an extension.

**Comment by chryzsh** [comment]
Great! Thanks for your sharp replies. I know have some idea about how this works. You talk about Eclipse, because the Archimatetool is based on Eclipse platform (RCP) am I right?

**Comment by jbsarrodie** [comment]
_You talk about Eclipse, because the Archimatetool is based on Eclipse platform (RCP) am I right?_

Yes.

Just an advice: don't underestimate the knowledge needed and workload implied by such project. It took me weeks to understand the basics behing Archi (from an Eclipse point of view) just to make some slight changes. And it's almost the same for every new feature I try to add. I really think the best is to start small (with some import export plugin and/or reports based on the HTML or Jasper one) for a "version 1" and then you'll have gain knowledge that will help you to create a real extension with better UI integration in a "version 2".

**Comment by jbsarrodie** [comment]
Another idea: the logic behing the "Validator" is to check some pre-defined rules. Maybe you could add your own rules to make sure models are compliant with your risk/security framework (properties sets...)

**Comment by Phillipus** [comment]
What JB says is true. For me, it took years of trial and error to figure out how the Eclipse framework works -  there's a frustrating lack of resources and documentation in the Eclipse world! I suggest to be clear about what you are trying to achieve before getting bogged down in Eclipse.

**Comment by chryzsh** [comment]
I will not be doing the actual developing, but I am responsible for charting what functionality we need and how it can be done and also what resources are necessary. I have just peeked at the source code and as you say, this will probably demand a lot from a developer. @thellef is my leader for this project. His vision is to be able to design models in Archi with focus on attributes for different security properties and reuirements and export them all the way to some web portal/dashboard for display, queries, etc.

I will however be looking into developing the front-end either by using the exported XML and handling it with XSLT or trying to expand upon your already given HTML report tool. I appreciate any recommendations and advice you can give us.
